layer/layer_store.go

```diff
@@ -419,11 +419,11 @@ func (ls *layerStore) Map() map[ChainID]Layer {
 func (ls *layerStore) deleteLayer(layer *roLayer, metadata *Metadata) error {
 	// Rename layer digest folder first so we detect orphan layer(s)
 	// if ls.driver.Remove fails
-	dir := ls.store.getLayerDirectory(layer.chainID)
+	var dir string
 	for {
 		dgst := digest.Digest(layer.chainID)
 		tmpID := fmt.Sprintf("%s-%s-removing", dgst.Hex(), stringid.GenerateRandomID())
-		dir := filepath.Join(ls.store.root, string(dgst.Algorithm()), tmpID)
+		dir = filepath.Join(ls.store.root, string(dgst.Algorithm()), tmpID)
 		err := os.Rename(ls.store.getLayerDirectory(layer.chainID), dir)
 		if os.IsExist(err) {
```

Comment:
Wondering if we can get into a situation where the `os.Rename()` succeeded once, but `ls.driver.Remove()` failed. In that case;

- layer directory was renamed from `aabbcc` to `aabbcc-<random-string>-removing`
- `ls.driver.Remove()` failed so the layer is still registered
- on a subsequent call, `deleteLayer()` (this function) will fail, because the expected directory (`aabbcc`) no longer was found (because it was renamed)

Should we either;

- rename `os.Rename` back to the original name if `ls.driver.Remove` failed?
- use a predictable name for the renamed directory (instead of using `stringid.GenerateRandomID()` ?

Reply:
The whole point of implementing layer garbase collector was handle correctly situation where `ls.driver.Remove()` fails and especially situation where layer is partly removed which causes very ugly issues as we cannot trust to layer integrity.

It is also tested by integration test
https://github.com/moby/moby/blob/ae0a878b863511d0455413b137d0f79b56e5d2f4/integration/image/remove_unix_test.go#L29-L113

Cleanup() function https://github.com/moby/moby/blob/a4a0429eec76dc2bb7d7649b8c851b2ec0f0beeb/layer/layer_store.go#L762-L781 will find those using layer _reference_ folder with name `aabbcc-<random-string>-removing` and only remove those after `ls.driver.Remove()` have succeed.

The reason why this PR is needed that we also remove that layer reference after `ls.driver.Remove()` have succeed (now only layer is removed but reference was left to disk).

What comes to `stringid.GenerateRandomID()` it was sequential number on some early version but I changed it to current format because requested change during review. On theory it should be possible to use cacheID instead of it but I'm not 100% if we can trust that it is always free.